### PR TITLE
[Flink-10079] [table] Support external sink table in the INSERT INTO clause 

### PR DESCRIPTION
**Pull request by tragicjun**
## What is the purpose of the change

In the documentation, there is a description: 

> Once registered in a TableEnvironment, all tables defined in a ExternalCatalog can be accessed from Table API or SQL queries by specifying their full path, such as catalog.database.table.

Currently, this is true only for source tables. For sink table (specified as **insertInto** in the Table API or SQL), the users have to explicitly register it even though it is defined in a registered ExternalCatalog, otherwise "No table was registered under the name XXX" TableException would be thrown. 

It would be better keep consistent for both source table and sink table, and the users would enjoy more convenient approach to inserting into sink tables. This pull request tries to automatically register sink table  if it is defined in any registered external catalogs.

## Brief change log

  - Try automatic registration of sink table in `TableEnvironment.insertInto()`.
  - Add unit test `ExternalCatalogInsertTest` to expose the issue.
  - Create a new external sink table in `CommonTestData.getInMemoryTestCatalog()`. 

## Verifying this change

This change added unit tests in `ExternalCatalogInsertTest` and can be verified accordingly.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zentol** [comment]
scalastyle violations:
```
10:30:04.747 [INFO] --- scalastyle-maven-plugin:1.0.0:check (default) @ flink-table_2.11 ---
error file=/home/travis/build/apache/flink/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala message=File line length exceeds 100 characters line=764
error file=/home/travis/build/apache/flink/flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/ExternalCatalogInsertTest.scala message=File line length exceeds 100 characters line=32
error file=/home/travis/build/apache/flink/flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/ExternalCatalogInsertTest.scala message=File line length exceeds 100 characters line=33
Saving to outputFile=/home/travis/build/apache/flink/flink-libraries/flink-table/target/scalastyle-output.xml
Processed 675 file(s)
```

**Comment by tragicjun** [comment]
@twalthr @suez1224 could you please review?

**Comment by tragicjun** [comment]
CC @zentol @fhueske

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala [hunk 1]
```diff
@@ -750,7 +751,28 @@ abstract class TableEnvironment(val config: TableConfig) {
     if (null == sinkTableName) throw TableException("Name of TableSink must not be null.")
     if (sinkTableName.isEmpty) throw TableException("Name of TableSink must not be empty.")
     if (!isRegistered(sinkTableName)) {
-      throw TableException(s"No table was registered under the name $sinkTableName.")
+      // try resolving and registering sink table from registered external catalogs
+      try {
+        val paths = sinkTableName.split("\\.")
+        if (paths.length > 1) {
```
Comment (walterddr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: I am not exactly sure but this seems problematic to me. is there a guarantee that the first element of the path will always be `catalog` name? This also depends on whether the path requires to be fully-qualify or not, right?

Comment (tragicjun) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: @walterddr thanks for reviewing. I assume full-path is required for both source table and sink table, as currently there is no "**USE** <db>" style of switching catalog/database supported. If fully qualified name is enforced, the first element of the path must be `catalog` name, right?

**Comment by tragicjun** [comment]
@walterddr @twalthr @fhueske could you please take a look, I thought it was a nice-to-have add-on in the Table API.

**Comment by tragicjun** [comment]
CC @zentol @xccui  would you please review on it?

**Comment by fhueske** [comment]
Hi @tragicjun, sorry for the late reply.
I'll have a look at this PR next week.

Best, Fabian

**Comment by tragicjun** [comment]
@fhueske would you have a look at it?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala [hunk 2]
```diff
@@ -750,7 +751,28 @@ abstract class TableEnvironment(val config: TableConfig) {
...
+          var externalCatalog = getRegisteredExternalCatalog(paths(0))
+          for (i <- 1 to (paths.length - 2)) {
+            externalCatalog = externalCatalog.getSubCatalog(paths(i))
+          }
+          val externalTable = externalCatalog.getTable(paths(paths.length - 1))
+          if (externalTable.isTableSink) {
+            registerTableSink(sinkTableName,
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: We should not register the sink table again in the root catalog.

**Comment by fhueske** [comment]
Please also rebase your PR on the current master.

Thank you,
Fabina

**Comment by tragicjun** [comment]
Thanks @fhueske. I take a closer look at the logic, looks like external catalogs are encapsulated as subschemas of the root schema.

The source tables are checked with subschemas during query validation, but sink tables are not checked accordingly. As per your suggestion, I've fixed `isRegistered()`, `getTable()`, `getRowType()` to check with subschemas, so that the full path of an external sink table are now supported in the `INSERT INTO` clause.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala [hunk 3]
```diff
@@ -820,20 +820,59 @@ abstract class TableEnvironment(val config: TableConfig) {
...
+      val (externalSchema, externalTableName) = resolveExternalTable(name)
+      if (externalSchema != null) {
+        isRegistered = externalSchema.getTableNames.contains(externalTableName)
+      }
+    }
+
+    return isRegistered
   }
 
   protected def getTable(name: String): org.apache.calcite.schema.Table = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: the method should return `Option[org.apache.calcite.schema.Table]`. We need to adjust `StreamTableEnvironment` and `BatchTableEnvironment` for this and not wrap the result of the `getTable()` calls another time in `Option`.

**Comment by tragicjun** [comment]
Thanks @fhueske for detailed comments. The suggestions look great to me, I've improved the PR based on them, please take a review again, thanks.

**Comment by tragicjun** [comment]
Thanks @fhueske , looks like `var` in scala is not recommended, sorry that I'm still getting used to scala style. I've replaced it with an recursive internal function. 

As for the nesting `match` blocks, I would like to keep it consistent with similar case in `StreamTableEnvironment.registerTableSinkInternal()` and `BatchTableEnvironment.registerTableSinkInternal()`, which I think is better than `if` statement. What do you think? Or please just improve it if there is more clean way. Thanks.

**Comment by fhueske** [comment]
Thanks @tragicjun.
I will merge this PR.

Best, Fabian

**Pull request closed without merging**